File: data_structure/doubly_linked_list.py
# https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_3_C

class Node:
    def __init__(self, key, prev, next):
        self.key = key
        self.prev = prev
        self.next = next


class DoublyLinkedList:
    def __init__(self):
        self.sentinel = Node(None, None, None)
        self.sentinel.prev = self.sentinel
        self.sentinel.next = self.sentinel

    def insert(self, key):
        node = Node(key, self.sentinel, self.sentinel.next)
        self.sentinel.next.prev = node
        self.sentinel.next = node

    def delete(self, key):
        node = self.sentinel.next

        while node is not self.sentinel:
            if node.key == key:
                node.prev.next = node.next
                node.next.prev = node.prev
                break
            node = node.next

    def deleteFirst(self):
        self.sentinel.next = self.sentinel.next.next
        self.sentinel.next.prev = self.sentinel

    def deleteLast(self):
        node = self.sentinel.prev

        node.prev.next = self.sentinel
        self.sentinel.prev = node.prev

    def print(self):
        node = self.sentinel.next

        buffer = []
        while node is not self.sentinel:
            buffer.append(str(node.key))
            node = node.next

        print(*buffer)


if __name__ == "__main__":
    n = int(input())

    commands = []

    for _ in range(n):
        command = input().split()
        if len(command) == 2:
            commands.append((command[0], int(command[1])))
        if len(command) == 1:
            commands.append((command[0], None))

    dll = DoublyLinkedList()

    for c in commands:
        if c[0] == 'insert':
            dll.insert(c[1])
        elif c[0] == 'delete':
            dll.delete(c[1])
        elif c[0] == 'deleteFirst':
            dll.deleteFirst()
        elif c[0] == 'deleteLast':
            dll.deleteLast()

    dll.print()

File: data_structure/doubly_linked_list.cpp
// https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_3_C

#include <iostream>
#include <vector>

using std::cin;
using std::cout;

struct Node {
    int key;
    Node *prev;
    Node *next;
};

class DoublyLinkedList {
    Node *sentinel;

public:
    DoublyLinkedList() {
        sentinel = new Node;
        sentinel->prev = sentinel;
        sentinel->next = sentinel;
    }

    void insert(int key) {
        Node *node = new Node;
        node->key = key;

        node->next = sentinel->next;
        sentinel->next->prev = node;

        sentinel->next = node;
        node->prev = sentinel;
    }

    void del(Node *node) {
        if (node == sentinel) {
            return;
        }
        node->prev->next = node->next;
        node->next->prev = node->prev;
        delete node;
    }

    void delByKey(int k) {
        Node *node = sentinel->next;

        while (node != sentinel && node->key != k) {
            node = node->next;
        }

        del(node);
    }

    void delFirst() { del(sentinel->next); }

    void delLast() { del(sentinel->prev); }

    void print() {
        Node *node = sentinel->next;

        bool is_first = true;
        while (true) {
            if (node == sentinel) {
                break;
            }
            if (is_first == false) {
                cout << " ";
            }
            cout << node->key;
            node = node->next;
            is_first = false;
        }
        cout << std::endl;
    }
};

int main() {
    DoublyLinkedList dll;

    cin.tie(nullptr);
    std::ios_base::sync_with_stdio(false);

    int n;
    cin >> n;

    std::string command;
    int key;
    for (int i = 0; i < n; i++) {
        cin >> command;

        if (command == "insert") {
            cin >> key;
            dll.insert(key);
        } else if (command == "delete") {
            cin >> key;
            dll.delByKey(key);
        } else if (command == "deleteFirst") {
            dll.delFirst();
        } else {
            dll.delLast();
        }
    }

    dll.print();
    return 0;
}